src/Microsoft.ML.Recommend/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,403 @@
+                _outputSchema = outputSchema;
+            }
+
+            public ISchema Schema => OutputSchema;
+
+            public ISchema OutputSchema
+            {
+                get { return _outputSchema; }
+            }
+
+            public Func<int, bool> GetDependencies(Func<int, bool> predicate)
+            {
+                for (int i = 0; i < OutputSchema.ColumnCount; i++)
+                {
+                    if (predicate(i))
+                        return col => (col == _xColIndex || col == _yColIndex);
+                }
+                return col => false;
+            }
+
+            public IEnumerable<KeyValuePair<RoleMappedSchema.ColumnRole, string>> GetInputColumnRoles()
+            {
+                yield return RecommendUtils.XKind.Bind(_xColName);
+                yield return RecommendUtils.YKind.Bind(_yColName);
+            }
+
+            private void CheckInputSchema(ISchema schema, int xCol, int yCol)
+            {
+                var type = schema.GetColumnType(xCol);
+                if (!type.Equals(_parent.InputXType))
+                {
+                    throw Contracts.ExceptParam(nameof(schema), "Input X type '{0}' incompatible with predictor X type '{1}'",
+                        type, _parent.InputXType);
+                }
+                type = schema.GetColumnType(yCol);
+                if (!type.Equals(_parent.InputYType))
+                {
+                    throw Contracts.ExceptParam(nameof(schema), "Input Y type '{0}' incompatible with predictor Y type '{1}'",
+                        type, _parent.InputYType);
+                }
+            }
+
+            private Delegate[] CreateGetter(IRow input, bool[] active)
+            {
+                Contracts.CheckValue(input, nameof(input));
+                Contracts.Assert(Utils.Size(active) == OutputSchema.ColumnCount);
+
+                var getters = new Delegate[1];
+                if (active[0])
+                {
+                    CheckInputSchema(input.Schema, _xColIndex, _yColIndex);
+                    var xGetter = input.GetGetter<uint>(_xColIndex);
+                    var yGetter = input.GetGetter<uint>(_yColIndex);
+                    getters[0] = _parent.GetGetter(xGetter, yGetter);
+                }
+                return getters;
+            }
+
+            public IRow GetRow(IRow input, Func<int, bool> predicate, out Action disposer)
+            {
+                var active = Utils.BuildArray(OutputSchema.ColumnCount, predicate);
+                var getters = CreateGetter(input, active);
+                disposer = null;
+                return new SimpleRow(OutputSchema, input, getters);
+            }
+
+            public ISchemaBindableMapper Bindable { get { return _parent; } }
+        }
+
+        public int UserFeaturesSize { get { return -1; } }
+        public KeyType UserIdType { get { return _inputYType; } }
+        public KeyType ItemIdType { get { return _inputXType; } }
+
```

Comment:
can these be grouped with the other properties above?

Reply:
We no longer have those fields.

---
In reply to: [225356944](https://github.com/dotnet/machinelearning/pull/1263#discussion_r225356944) [](ancestors = 225356944)